static/app/components/tours/tourContext.tsx

```diff
@@ -0,0 +1,158 @@
+import {type Dispatch, type Reducer, useCallback, useState} from 'react';
+import {useReducer} from 'react';
+
+export type TourEnumType = string | number;
+
+export interface TourStep<T extends TourEnumType> {
+  /**
+   * The element to focus on when the tour step is active.
+   * This is usually set within the TourElement component, which wraps the focused element.
+   */
+  element: HTMLElement | null;
+  /**
+   * Unique ID for the tour step.
+   */
+  id: T;
+}
+
+type TourStartAction<T extends TourEnumType> = {
+  type: 'START_TOUR';
+  stepId?: T;
+};
+type TourNextStepAction = {
+  type: 'NEXT_STEP';
+};
+type TourPreviousStepAction = {
+  type: 'PREVIOUS_STEP';
+};
+type TourEndAction = {
+  type: 'END_TOUR';
+};
+type TourRegisterStepAction<T extends TourEnumType> = {
+  step: TourStep<T>;
+  type: 'REGISTER_STEP';
+};
+
+export type TourAction<T extends TourEnumType> =
+  | TourStartAction<T>
+  | TourNextStepAction
+  | TourPreviousStepAction
+  | TourEndAction
+  | TourRegisterStepAction<T>;
+
+export interface TourState<T extends TourEnumType> {
+  /**
+   * The current active tour step. If this is null, the tour is not active.
+   */
+  currentStep: TourStep<T> | null;
+  /**
+   * Whether the tour is available to the user. Should be set by flags or other conditions.
+   */
+  isAvailable: boolean;
+  /**
+   * The ordered step IDs. Declared once when the provider is initialized.
+   */
+  orderedStepIds: readonly T[];
+}
+
+type TourRegistry<T extends TourEnumType> = {
+  [key in T]: TourStep<T> | null;
+};
+
+export function useTourReducer<T extends TourEnumType>(
+  initialState: TourState<T>
+): TourContextType<T> {
+  const {orderedStepIds} = initialState;
+  const [registry, setRegistry] = useState<TourRegistry<T>>(
```

Comment:
Does the registry need to be state? I'd prefer to store it in a ref, otherwise every element that registers itself is is going to cause a rerender

Reply:
this is a great catch, thank you!